File: DownToZeroII.py
from math import sqrt
v = [0]*1001010
n = 0
def bfs(c,q):
    global v
    m = c[0]
    for i in range(2,int(round(sqrt(m)))):
        if m%i == 0:
            nt = max(i, m//i)
            if v[nt]:
                v[nt] = 1
                q.append([nt, c[1] + 1])    
    if m and not v[m - 1]:
        v[m - 1] = 1
        q.append([m - 1, c[1] + 1])
    return q
    
t = int(input().strip())
while t>0:
    n = int(input().strip())
    q = []
    ans = 0
    q.append([n, 0])
    while q:
        c = q[0]
        q.pop()
        if c[0] == 0:
            ans = c[1]
            break       
        bfs(c, q)    
    print("%i\n" %(ans))
    t-=1
    

File: DownToZeroII.cpp
//#include <cmath>
//#include <cstring>
//#include <cstdio>
//#include <vector>
//#include <queue>
//#include <iostream>
//#include <algorithm>
//using namespace std;
//const int ln = 1001010;
//bool v[ln];
//int n;
//void start(){
//    memset(v, 0, sizeof(v));
//}
//
//void calcule(pair<int,int> &c, queue<pair<int,int>> &q){
//    int m = c.first;
//    for(int i = 2; i <= sqrt(m); ++ i){
//        if(m%i == 0){
//            int nt = max(i, m/i);
//            if(!v[nt]){
//                v[nt] = 1;
//                q.push(make_pair(nt, c.second + 1));
//            }
//        }
//    }
//    if(m && !v[m - 1]){
//        v[m - 1] = 1;
//        q.push(make_pair(m - 1, c.second + 1));
//    }
//}
//
//int main() {
//    int T;
//    scanf("%d",&T);
//    while(T --){
//        scanf("%d",&n);
//        start();
//        queue<pair<int,int>> q;
//        int ans = 0;
//        q.push(make_pair(n, 0));
//        while(!q.empty()){
//            pair<int,int> cur = q.front();
//            q.pop();
//            if(cur.first == 0){
//                ans = cur.second;
//                break;
//            }
//            calcule(cur, q);
//        }
//        printf("%d\n",ans);
//    }
//    return 0;
//}
